parser-typechecker/src/Unison/Runtime/ANF.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# Language OverloadedStrings #-}
+{-# Language PatternSynonyms #-}
+{-# Language ScopedTypeVariables #-}
+
+module Unison.Runtime.ANF (optimize, fromTerm, fromTerm', term) where
+
+import Data.Foldable hiding (and,or)
+import Data.List hiding (and,or)
+import Prelude hiding (abs,and,or)
+import Unison.Term
+import Unison.Var (Var)
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Unison.ABT as ABT
+import qualified Unison.Term as Term
+import qualified Unison.Var as Var
+import Unison.Typechecker.Components (minimize')
+
+newtype ANF v a = ANF_ { term :: Term.AnnotatedTerm v a }
+
+-- Replace all lambdas with free variables with closed lambdas.
+-- Works by adding a parameter for each free variable. These
+-- synthetic parameters are added before the existing lambda params.
+-- For example, `(x -> x + y + z)` becomes `(y z x -> x + y + z) y z`.
+-- As this replacement has the same type as the original lambda, it
+-- can be done as a purely local transformation, without updating any
+-- call sites of the lambda.
+--
+-- The transformation is shallow and doesn't transform the body of
+-- lambdas it finds inside of `t`.
+lambdaLift :: (Var v, Semigroup a) => (v -> v) -> AnnotatedTerm v a -> AnnotatedTerm v a
+lambdaLift liftVar t = ABT.visitPure go t where
+  go t@(LamsNamed' vs body) = Just $ let
+    fvs = ABT.freeVars t
+    fvsLifted = [ (v, liftVar v) | v <- toList fvs ]
+    a = ABT.annotation t
+    subs = [(v, var a v') | (v,v') <- fvsLifted ]
+    in if Set.null fvs then lam' a vs body -- `lambdaLift body` would make transform deep
+       else apps' (lam' a (map snd fvsLifted ++ vs) (ABT.substs subs body))
+                  (snd <$> subs)
+  go _ = Nothing
+
+optimize :: forall a v . (Semigroup a, Var v) => AnnotatedTerm v a -> AnnotatedTerm v a
+optimize t = go t where
+  ann = ABT.annotation
+  go (Let1' b body) | canSubstLet b body = go (ABT.bind body b)
+  go e@(App' f arg) = case go f of
+    Lam' f -> go (ABT.bind f arg)
+    f -> app (ann e) f (go arg)
+  go (If' (Boolean' False) _ f) = go f
+  go (If' (Boolean' True) t _) = go t
+  -- todo: can simplify match expressions
+  go e@(ABT.Var' _) = e
+  go e@(ABT.Tm' f) = case e of
+    Lam' _ -> e -- optimization is shallow - don't descend into lambdas
+    _ -> ABT.tm' (ann e) (go <$> f)
+  go e@(ABT.out -> ABT.Cycle body) = ABT.cycle' (ann e) (go body)
+  go e@(ABT.out -> ABT.Abs v body) = ABT.abs' (ann e) v (go body)
+  go e = e
+
+  -- test for whether an expression `let x = y in body` can be
+  -- reduced by substituting `y` into `body`. We only substitute
+  -- when `y` is a variable or a primitive, otherwise this might
+  -- end up duplicating evaluation or changing the order that
+  -- effects are evaluated
+  canSubstLet (Var' _) _body = True
```

Comment:
Since leaves are always pure primitives can Thisbe replaced be a guard `isLeaf e`?

Reply:
@noahhaasis yeah that sounds right - `canSubstLet` is just `isLeaf`. 

I was thinking we could also do the substitution if the expression is pure (requires no abilities) and only occurs once in the body of the let, but we don't have an easy way to check for purity here.